### ISSUE TITLE
Spring boot cannot find velocity toolbox.xml inside jar file

### ISSUE THREAD

**Issue by aluisiodsv** [spring-projects/spring-boot#2912]
When create a simple spring boot application with spring-boot-starter-velocity for velocity templates I was unable to load toolbox.xml file even if I set the spring.velocity.toolbox-config-location property.
When running the application this is the log output when the logging.level.org.apache.velocity was set to DEBUG:
```
2015-05-04 16:57:53.954 DEBUG 27289 --- [nio-9393-exec-1] o.a.v.t.v.servlet.ServletToolboxManager  : No toolbox was found at '/toolbox.xml'
2015-05-04 16:57:53.960 ERROR 27289 --- [nio-9393-exec-1] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.NullPointerException] with root cause
java.lang.NullPointerException: null
        at org.springframework.web.servlet.view.velocity.VelocityToolboxView.createVelocityContext(VelocityToolboxView.java:111)
        at org.springframework.web.servlet.view.velocity.VelocityView.renderMergedTemplateModel(VelocityView.java:288)
        at org.springframework.web.servlet.view.AbstractTemplateView.renderMergedOutputModel(AbstractTemplateView.java:167)
        at org.springframework.web.servlet.view.AbstractView.render(AbstractView.java:303)
        at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1244)
        at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:1027)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:971)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:966)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:857)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:618)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:842)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.springframework.web.filter.ttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:85)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:516)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1086)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:659)
        at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.process(Http11NioProtocol.java:223)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1558)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1515)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.apache.tomcat.util.threads.TaskThrepingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:745)
2015-05-04 16:57:53.967 DEBUG 27289 --- [nio-9393-exec-1] org.apache.velocity                      : ResourceManager : found error.vm with loader org.apache.velocity.runtime.resource.loader.FileResourceLoader
```
Unfortunately I did not find sufficient documentation in this case.

**Comment by wilkinsona** [comment]
Velocity attempts to load its toolbox config file using the `ServletContext`. How are you running your application and what servlet container are you using? Perhaps you can provide a small project that reproduces the problem?

**Comment by aluisiodsv** [comment]
Off course. It is just a simple project with spring-boot-starter-velocity who use Tomcat as servlet container and is executable jar.
https://drive.google.com/file/d/0B9PUSyNx7RbFLTNJR19sV085RTg/view?usp=sharing

**Comment by wilkinsona** [comment]
Thanks

**Comment by aluisiodsv** [comment]
Any feedback?

**Comment by wilkinsona** [comment]
Not yet, sorry. I've been out on paternity leave. I'll try to get back to this soon.

**Comment by philwebb** [comment]
@aluisiodsv 

For some reason I can't unzip the archive.

```
Archive:  velocity-boot.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of velocity-boot.zip or
        velocity-boot.zip.zip, and cannot find velocity-boot.zip.ZIP, period.
```

Could you upload it again or put it on GitHub?

**Comment by aluisiodsv** [comment]
OMG. Google drive lost my data. oO
I will post again. Sry.

**Comment by aluisiodsv** [comment]
Better here in github
https://github.com/aluisiodsv/velocity-boot.git

**Comment by philwebb** [comment]
The `ServletToolboxManager` is trying to load the `toolbox.xml` from the `ServletContext` using `getResourceAsStream`. Unfortunately this wont work for an embedded jar. 

We can probably create a new `VelocityToolboxView` that loads the xml in a different way.

**Comment by aluisiodsv** [comment]
Thanks.

**Comment by yglodt** [comment]
Not sure if it's related, but it looks like Velocity can also not find the templates when it's running inside a Spring Boot application which in turn runs as a war file. While the project is running in Eclipse it works though.

**Comment by wilkinsona** [comment]
@yglodt I've just converted `spring-boot-sample-web-velocity` to a war file. It works fine both when run with `java -jar` and when deployed to Tomcat. Please open a new issue if you can provide a sample project that shows it not working.

**Comment by yglodt** [comment]
@wilkinsona Thanks for you feedback. Just one more comment. I do not use Velocity as jsp-replacement, but for email content, and call it like this:

`VelocityEngineUtils.mergeTemplateIntoString(velocityEngine, "/mailing.html", StandardCharsets.UTF_8.name(), model);
`
Maybe that makes a difference ...

**Comment by wilkinsona** [comment]
@yglodt That appears to work too, however you haven't provided enough information for me to be sure. You said above that "Velocity can also not find the templates", but this issue is about Velocity not being able to find it's toolbox.xml configuration file. As I said above, please open a new issue if you believe you have found a problem. If you do so, please provide a small sample project that reproduces the problem so that we can be certain that we're talking about exactly the same thing.

**Comment by icanfly** [comment]
I met the some problem too.

**Comment by wilkinsona** [comment]
@icanfly We believe that this issue have been fixed in Spring Boot 1.2.5.RELEASE (and later). If you believe that is not the case, please open a new issue with a small sample project that reproduces the problem.

**Comment by huadi** [comment]
Spring Boot 1.4.3.RELEASE, cannot find XML config too.
The `VelocityToolboxView` called `ServletToolboxManager.getInstance()` which still use `servletContext.getResourceAsStream()` method to get `toolbox.xml`, and cannot load inside jar.
Is there something I miss? How to fixed this problem by 
> We can probably create a new `VelocityToolboxView` that loads the xml in a different way.
on 16 Jun 2015
